web/interceptor/user_auth.go

```diff
@@ -11,36 +11,76 @@ import (
 	"github.com/quickfeed/quickfeed/web/auth"
 )
 
+type UserInterceptor struct {
+	tm     *auth.TokenManager
+	logger *zap.SugaredLogger
+}
+
+func NewUserInterceptor(logger *zap.SugaredLogger, tm *auth.TokenManager) *UserInterceptor {
+	return &UserInterceptor{
+		tm:     tm,
+		logger: logger,
+	}
+}
+
+func (u *UserInterceptor) WrapStreamingHandler(next connect.StreamingHandlerFunc) connect.StreamingHandlerFunc {
+	return connect.StreamingHandlerFunc(func(ctx context.Context, conn connect.StreamingHandlerConn) error {
+		cookie := conn.RequestHeader().Get(auth.Cookie)
+		claims, err := u.tm.GetClaims(cookie)
+		if err != nil {
+			return connect.NewError(connect.CodeUnauthenticated, fmt.Errorf("failed to extract JWT claims from session cookie: %w", err))
+		}
+
+		var updatedCookie *http.Cookie
+		if u.tm.UpdateRequired(claims) {
+			u.logger.Debug("Updating cookie for user ", claims.UserID)
+			updatedCookie, err = u.tm.UpdateCookie(claims)
+			if err != nil {
+				return connect.NewError(connect.CodeUnauthenticated, fmt.Errorf("failed to update session cookie: %w", err))
+			}
+		}
+		if updatedCookie != nil {
+			conn.ResponseHeader().Set(auth.SetCookie, updatedCookie.String())
+		}
+
+		return next(claims.ClaimsContext(ctx), conn)
+	})
+}
+
+func (*UserInterceptor) WrapStreamingClient(next connect.StreamingClientFunc) connect.StreamingClientFunc {
+	return connect.StreamingClientFunc(func(ctx context.Context, spec connect.Spec) connect.StreamingClientConn {
+		return next(ctx, spec)
```

Comment:
I guess we don't support streaming from clients, but should we instead return an error? I'm not sure how, though, since the signature doesn't have an error return type.

Moreover, I guess the same applies to all interceptors for streaming clients.